pkg/scheduler/algorithm/priorities/metadata.go

```diff
@@ -105,37 +105,52 @@ func getFirstServiceSelector(pod *v1.Pod, sl corelisters.ServiceLister) (firstSe
 	return nil
 }
 
-// getSelectors returns selectors of services, RCs and RSs matching the given pod.
-func getSelectors(pod *v1.Pod, sl corelisters.ServiceLister, cl corelisters.ReplicationControllerLister, rsl appslisters.ReplicaSetLister, ssl appslisters.StatefulSetLister) []labels.Selector {
-	var selectors []labels.Selector
+// getSelector returns a selector for the services, RCs, RSs, and SSs matching the given pod.
+func getSelector(pod *v1.Pod, sl corelisters.ServiceLister, cl corelisters.ReplicationControllerLister, rsl appslisters.ReplicaSetLister, ssl appslisters.StatefulSetLister) labels.Selector {
+	labelSet := make(labels.Set)
 
 	if services, err := sl.GetPodServices(pod); err == nil {
 		for _, service := range services {
-			selectors = append(selectors, labels.SelectorFromSet(service.Spec.Selector))
+			if labels.Conflicts(labelSet, service.Spec.Selector) {
```

Comment:
how is it possible that the pod belongs to two conflicting services? if this is something that we should not expect, I suggest we remove the check to simplify the code, add a comment that we ignore conflicts and silently merge the two label sets according to labels.Merge semantics since we can't do anything about it.

Reply:
Done

Reply:
Exactly, since the given pod matches, there shouldn't be any conflicts. I didn't think about that

Reply:
Apologize for the late review (I actually put a similar comment, but on a commit instead of PR files, then it's erased by a rebase...), the key difference here is: inside Kubernetes, there are 2 kinds of selector:

- simple selector, a.k.a LabelSelector, which is represented as a `map[string]string` format. For this kind of selector, there is no chance they can conflict as by given a key, the matched value is also fixed.
- complex selector,a.k.a Requirement, which is represented as `key <operator> [values] format`. <operator> can be "In", "NotIn", "Exists", etc. [values] is a string array. In this case, two selectors may not conflict either, and we're simply `appending` them.

---

As we're here, an open question comes into my mind, can we pre-merge them and then check if the merged selector matches the Pod labels? This _may_ make the logic faster.

For example, we have 2 "complex selectors":
    - selector A: foo In [X, Y]
    - selector B: foo NotIn [Y, Z]
 
Pre-merge can make a merged selector: `foo In [X]`, instead of `[foo In [X, Y], foo NotIn [Y, Z]]`. But right now `Requirement` doesn't have "Merged" implemented.

Reply:
I thought about it, but I can imagine merging ending up being very complex.
However, with the current implementation, if a pod matches a service and a replication controller, we can have something like `[foo = bar, foo = bar]`. We could scope merging to only handle these obvious cases.
But this is still a significant effort which we can leave for later optimizations.

Reply:
Agree, we should keep an eye on it as from #85157, it seems we can optimize the "matching logic" to gain perf boosting.